File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    const oa = {};
    const ob = {};
    for (let i = 0; i < t.length; i++) {
        oa[t[i]] = (oa[t[i]] || 0) + 1;
    }
    const res = new Map();
    const st = new Set();
    const rc = Object.keys(oa).length;
    let i = 0;
    for (let j = 0; j < s.length; j++) {
        ob[s[j]] = (ob[s[j]] || 0) + 1;
        if (s[j] in oa && ob[s[j]] === oa[s[j]]) {
            st.add(s[j]);
        }
        if (st.size === rc) {
            while (i < s.length && (!oa[s[i]] || ob[s[i]] > oa[s[i]])) {
                ob[s[i]] -= 1;
                i += 1;
            }
            res.set((j - i), [i, j]);
        }
    }
    if (res.size) {
        const minVal = Math.min(...res.keys());
        const [i, j] = res.get(minVal);
        return s.slice(i, j + 1);
    }
    return '';
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        oa = defaultdict(int)
        ob = defaultdict(int)
        st = set()
        res = {}
        for i in range(len(t)):
            oa[t[i]] = oa.get(t[i], 0) + 1

        i = 0
        for j in range(len(s)):
            ob[s[j]] = ob.get(s[j], 0) + 1

            while i < len(s) and s[i] in oa and s[i] in st and ob[s[i]] < oa[s[i]]:
                    st.remove(s[i])
            while i < len(s) and s[i] not in oa:
                ob[s[i]] -= 1
                i += 1

            if s[j] in oa and ob[s[j]] == oa[s[j]]:
                st.add(s[j])
            if len(st) == len(oa):
                while s[i] not in oa or ob[s[i]] > oa[s[i]]:
                    ob[s[i]] -= 1
                    i += 1
                res[j - i] = (i, j)
        print(res)
        if len(res):
            i, j = res[min(res.keys())]
            return s[i: j + 1]
        return ''

